libvast/builtins/formats/json.cpp

```diff
@@ -0,0 +1,356 @@
+      auto result = val.get_int64();
+      if (result.error())
+        return result.error();
+      pusher.add(result.value_unsafe());
+      return {};
+    }
+    case simdjson::ondemand::number_type::unsigned_integer: {
+      auto result = val.get_uint64();
+      if (result.error())
+        return result.error();
+      pusher.add(result.value_unsafe());
+      return {};
+    }
+  }
+  die("unhandled json number type in switch statement");
+}
+
+parse_result parse_string(simdjson::ondemand::value val, auto& pusher) {
+  auto maybe_str = val.get_string();
+  if (maybe_str.error())
+    return maybe_str.error();
+  auto str = val.get_string().value_unsafe();
+  using namespace parser_literals;
+  static constexpr auto parser
+    = parsers::time | parsers::duration | parsers::net | parsers::ip
+      | as<caf::none_t>("nil"_p) | as<caf::none_t>(parsers::ch<'_'>);
+  data result;
+  if (parser(str, result)) {
+    pusher.add(make_view(result));
+    return {};
+  }
+  // Take the input as-is if nothing worked.
+  pusher.add(str);
+  return {};
+}
+
+parse_result
+parse_array(simdjson::ondemand::value val, auto& pusher, size_t depth) {
+  auto result = val.get_array();
+  if (result.error())
+    return result.error();
+  auto lst = result.value_unsafe();
+  auto list = pusher.push_list();
+  for (auto element : lst) {
+    if (element.error())
+      return element.error();
+    if (auto parsed_element
+        = parse_impl(element.value_unsafe(), list, depth + 1))
+      return parsed_element;
+  }
+  return {};
+}
+
+parse_result
+parse_object(simdjson::ondemand::value val, auto& pusher, size_t depth) {
+  auto result = val.get_object();
+  if (result.error())
+    return result.error();
+  auto obj = result.value_unsafe();
+  auto record = pusher.push_record();
+  for (auto pair : obj) {
+    if (pair.error())
+      return pair.error();
+    auto key = pair.unescaped_key();
+    if (key.error())
+      return key.error();
+    auto val = pair.value();
+    if (val.error())
+      return val.error();
+    auto field = record.push_field(key.value_unsafe());
+    if (auto parse_result = parse_impl(val.value_unsafe(), field, depth + 1))
+      return parse_result;
+  }
+  return {};
+}
+
+parse_result
+parse_impl(simdjson::ondemand::value val, auto& pusher, size_t depth) {
+  if (depth > defaults::max_recursion)
+    die("nesting too deep in json_parser parse");
+  auto type = val.type();
+  if (type.error())
+    return type.error();
+  switch (val.type().value_unsafe()) {
+    case simdjson::ondemand::json_type::null:
+      return {};
+    case simdjson::ondemand::json_type::number:
+      return parse_number(val, pusher);
+    case simdjson::ondemand::json_type::boolean: {
+      auto result = val.get_bool();
+      if (result.error())
+        return result.error();
+      pusher.add(result.value_unsafe());
+      return {};
+    }
+    case simdjson::ondemand::json_type::string:
+      return parse_string(val, pusher);
```

Comment:
This is because the type() function used in switch statement is very optimistic. It only checks the first character.
https://github.com/simdjson/simdjson/blob/master/singleheader/simdjson.h#L29098

I think we can have similar examples that yield a correct type but raises an error when we try to get a value like in the double example above

Reply:
Oh wow, that's crazy. Can you link to https://github.com/simdjson/simdjson/blob/f435fddda148bec9e798dd3bdb1988af569d40af/singleheader/simdjson.h#L29098-L29117 (permalink) in the code for the future reader?